topics/Padroes_de_Codigo.md

```diff
@@ -0,0 +1,115 @@
+tem requisitos diferentes.
+
+<!-- toc -->
+
+## Imutabilidade
+
+Imutabilidade é um conceito muito presente em linguagens funcionais, que buscam
+garantir que uma função sempre retornará o mesmo valor, possibilitando até que
+a funcionalidade alí definida possa ser provada matemáticamente. Para que isso
+seja verdade, é preciso garantir que funções não mantenham estado e não tenham
+valores variáveis (ou seja, sejam imutáveis).
+
+Porém, esse conceito tem evoluido além de linguagens funcionais, pois
+imutabilidade evita também que a sua aplicação retorne informações conflitantes,
+causadas por concorrência, que fazem a aplicação impossível de debugar.
+
+### Recursos
+
+* [[Artigo] Objects Should be Immutable](http://www.yegor256.com/2014/06/09/objects-should-be-immutable.html)
+  :uk:
+
+## SOLID
+
+SOLID é um acrônimo que engloba os cinco princípios básicos de orientação a
+objetos. Existem variações, mas em geral eles se resumem a:
+
+* *Single responsibility principle* (princípio de responsabilidade única)
+  O principio de responsabilidade única define que uma classe ou método deveria
+  existir somente para implementar uma funcionalidade, mas que essa
+  responsabilidade seja completamente encapsulada pela entidade.
+* *Open/close principle* (princípio aberto/fechado)
+  O princípio aberto/fechado dita que entidades de software deveriam ser abertas
+  a extensão, mas fechadas a modificação. A ideia é que essas entidades
+  (classes, objetos, metodos...) possam ter suas funcionalidades extendidas,
+  mas não seu código modificado.
+* *Liskov substitution principle* (princípio da substituição de Liskov)
+  Esse princípio visa definir o conceito de subtipo, garantindo que subtipos
+  mantém as propriedades defininas no tipo original. Esse princípio afirma que
+  se S é um subtipo do objeto T, qualquer objeto de tipo T pode ser substituído
+  por objetos de tipo S.
+* *Interface segregation principle* (princípio de separação de interfaces)
+  ISP diz que nenhum cliente deveria dependender de metodos que ele não vai
+  usar. Promove interfaces menores, desacopladas, e expressivas.
+* *Dependency inversion principle* (princípio de inversão de depêndencias)
+  Esse príncipio define que módulos de alto-nível não dependem de módulos de
+  baixo-nível, mas que os dois dependem de abstrações. Além disso, abstrações
+  não deveriam depender de detalhes de implementação, mas o contrário.
+
+### Recursos
+
+* [[Wikipedia] SOLID (object-oriented design)](https://en.wikipedia.org/wiki/SOLID_%28object-oriented_design%29)
+  :uk:
+
+## Encapsulamento
+
+Quando Alan Kay criou o SmallTalk por volta de 1970,
+ele acreditava que cada objeto era uma representação de todos os recursos
+do computador, sendo, dessa forma, completamente independente. É nesse
+contexto que se formou a ideia de encapsulamento. Os seus componentes devem
+ser simples o suficiente para poderem ser transportados por toda a aplicação,
+sem perderem suas propriedades.
```

Comment:
me lembrou disso: http://rylev.github.io/words/blog/2013/10/03/erlang-is-the-most-object-oriented-language/ :P